src/tests/ftest/container/per_server_fault_domain.py

```diff
@@ -0,0 +1,330 @@
+        non_svc_ranks.remove(stop_rank_1)
+
+        # Now the list contains one or two elements. Select the first element and find the
+        # other rank that's on the same node as before.
+        stop_rank_3 = non_svc_ranks[0]
+        stop_rank_4 = None
+        for rank in range(8):
+            if rank != stop_rank_3:
+                if rank_to_ip[rank] == rank_to_ip[stop_rank_3]:
+                    stop_rank_4 = rank
+                    break
+        self.log.info("Stop rank 3 = %s; 4 = %s", stop_rank_3, stop_rank_4)
+
+        ranks_to_stop = f"{stop_rank_1},{stop_rank_2},{stop_rank_3},{stop_rank_4}"
+        self.log.info("Ranks to stop = %s", ranks_to_stop)
+
+        properties = self.params.get("rf_2", "/run/cont_property/*")
+
+        # Run step 2 to 8.
+        self.verify_per_server_fault(
+            create_pool=False, ranks_to_stop=ranks_to_stop, properties=properties,
+            expected_status="HEALTHY")
+
+    @skipForTicket("DAOS-12394")
+    def test_rf2_unclean(self):
+        """Jira ID: DAOS-11200
+
+        Test Description:
+        1. Determine the ranks to stop. Three ranks in three nodes. Select up to two
+        service ranks.
+        2. Create a container with rf_lvl:2,rf:2
+        3. Run IOR to write some data to the container.
+        4. Stop the ranks.
+        5. Wait for rebuild to finish.
+        6. Get container's health status.
+        7. Restart the stopped servers for cleanup.
+        8. Verify container's Health status is HEALTHY.
+
+        :avocado: tags=all,full_regression
+        :avocado: tags=hw,medium
+        :avocado: tags=container
+        :avocado: tags=per_server_fault_domain,rf2_unclean
+        """
+        # 1. Determine the ranks to stop; three ranks in three nodes. We can select up to
+        # two ranks from service ranks. If we stop more than two service ranks, many of
+        # the dmg commands will hang.
+        # Create a pool to determine the service ranks.
+        self.add_pool()
+        self.log.info("Pool service ranks = %s", self.pool.svc_ranks)
+
+        # Create the list of non-service ranks. Assume there are 5 service ranks and rank
+        # numbering is consecutive.
+        non_svc_ranks = []
+        for rank in range(8):
+            if rank not in self.pool.svc_ranks:
+                non_svc_ranks.append(rank)
+        self.log.info("non_svc_ranks = %s", non_svc_ranks)
+
+        # Create rank to IP dictionary.
+        rank_to_ip = self.create_rank_to_ip()
+
+        # Prepare IP set, which contains the IP of selected ranks to stop.
+        stop_rank_ip = set()
+
+        # Select first element in the list (arbitrary).
+        non_svc_rank = non_svc_ranks[0]
+        stop_rank_ip.add(rank_to_ip[non_svc_rank])
+        ranks_to_stop = []
+        ranks_to_stop.append(non_svc_rank)
+
+        # Iterate ranks and select the other two using rank to IP dictionary and IP set.
+        for rank in range(8):
```

Comment:
Instead of hardcoding 8, is it easy enough to make this adjustable?

Reply:
Updated to calculate the number of ranks (8) from node count and engine/node count. However, we'll most likely stick with 8 because if we use 3 nodes, this test would work, but another test wouldn't. The tests will work with more than 4 nodes, but then we'll have to migrate to HW Large and we want to avoid that unless it's absolutely necessary.